tensorflow_quantum/core/ops/tfq_simulate_expectation_op.cc

```diff
@@ -96,9 +97,43 @@ class TfqSimulateExpectationOp : public tensorflow::OpKernel {
     context->device()->tensorflow_cpu_worker_threads()->workers->ParallelFor(
         programs.size(), num_cycles, construct_f);
 
+    int max_num_qubits = 0;
+    for (const int num : num_qubits) {
+      max_num_qubits = std::max(max_num_qubits, num);
+    }
+
+    // Cross reference with standard google cloud compute instances
+    // Memory ~= 2 * num_threads * (2 * 64 * 2 ** num_qubits in circuits)
+    // e2s2 = 2 CPU, 8GB -> Can safely do 25 since Memory = 4GB
+    // e2s4 = 4 CPU, 16GB -> Can safely do 25 since Memory = 8GB
+    // ...
+    if (max_num_qubits < 26) {
+      ComputeSmall(num_qubits, max_num_qubits, fused_circuits, pauli_sums,
+                   context, &output_tensor);
+    } else {
+      ComputeLarge(num_qubits, fused_circuits, pauli_sums, context,
+                   &output_tensor);
+    }
+
+    // just to be on the safe side.
+    qsim_circuits.clear();
+    fused_circuits.clear();
+    num_qubits.clear();
+    maps.clear();
+    pauli_sums.clear();
+    programs.clear();
+  }
+
+ private:
+  void ComputeLarge(
+      const std::vector<int>& num_qubits,
+      const std::vector<std::vector<qsim::GateFused<QsimGate>>>& fused_circuits,
```

Comment:
I remember that it is redefined as QsimCircuit.

Reply:
QsimCircuit was for the non-fused circuits, when dealing with the fused circuits we have to define it in terms of QsimGates.